### ISSUE TITLE
Feature Request: Support for cross-domain downloads

### ISSUE THREAD

**Issue by aarongustafson** [w3c/webappsec-csp#323]
In markup, an author can make a link trigger a download:

```html
<a href="some-resource.ext" download>Download this file</a>
```

In some instances, the resource may reside on another domain (e.g., a CDN). If this is the case, for security reasons, [some browsers ignore the `download` attribute](https://bugs.chromium.org/p/chromium/issues/detail?id=714373), reducing it to a standard link.

If we added a "download" instruction to CSP, we could inform the browser of trusted domains that we would like to allow downloads from (just as we do for everything else).

Notes:

* It is possible to add a `Content-Disposition` header on the CDN to force download, but this would not allow for, say, an image to be displayed in the browser *and* downloaded via a link from the same URL. The domain/CDN would have to support a dynamic `Content-Disposition` header based on a query string (or similar).
* There are other ways to bypass this restriction using JavaScript, but they are not nearly as straightforward and declarative as `a[download]`.

**Comment by annevk** [comment]
CSP doesn't seem like the right solution here, since that's still in the "attacker's" control. You want something like CORS I suppose.

**Comment by aarongustafson** [comment]
@annevk Originally that was my thought as well, but the `a[download]` use case is much more similar to Origin A allowing content from Origin B to be embedded in a page (like frame src, etc.) than it is to Origin B controlling whether Origin A can display its content (which is the purview of CORS). Essentially, it’s a matter of Origin A trusting Origin B not to be a nefarious player.

Scenarios:

* Origin A had a CDN (Origin B) and trusts it as an image source (CSP). It also wants to enable download of said images via a link. Just as with image display, this requires trust and Origin A trusts Origin B (because it has some measure of control over it as a CDN).
* Origin C supports user generated content. In order to protect its users, it only allows embedding of media that is uploaded to its system (where it has security checks place to prevent malicious code injection) via CSP. If a user includes a download link to Origin D as part of their content, the `download` attribute should be ignored if it is not trusted by Origin C.

I’m not sure what you mean by this:

> CSP doesn't seem like the right solution here, since that's still in the "attacker's" control

What does "that" refer to?

**Comment by annevk** [comment]
Origin B might not be okay with its resources being downloaded, hence the need for some kind of explicit consent. (And why it is restricted to begin with.)

**Comment by aarongustafson** [comment]
> Origin B might not be okay with its resources being downloaded, hence the need for some kind of explicit consent. (And why it is restricted to begin with.)

True. It seems like something that makes sense for a combination of CORS and CSP because it requires trust on both ends. And in the event that a single entity owns both domains—which is a very common  case in primary domain + CDN scenarios—the ability to explicitly state "I trust Origin B and want to enable users to download from it directly from links on my page" is equally important. 

To get the desired behavior (click to download) with the current blocking behavior in Chrome, you have to add a ton of hacky JavaScript which further burdens the user’s experience or you have to use a CDN that supports dynamic headers *and* you have to use two distinct URIs to trigger the two behaviors: show & download.

This is very confusing to teach (and remember) because the behavior is different between origins. If a CSP instruction allowed you to treat another origin—that you trust…it should go without saying—equally, I think that’s a win all around and it’s consistent with how CSP is used today: Devs are already used to doing this for video embeds & such.

**Comment by aarongustafson** [comment]
Whoops, hit the wrong button.

**Comment by annevk** [comment]
I don't see why you need to declare trust in Origin B other than by including links to them in your page.

**Comment by aarongustafson** [comment]
@annevk There’s two sides to the equation and each has risks. There are probably more, but these are the one’s currently on my mind:

| Role | Risk | Control |
|---|---|---|
| Web page linking to a cross-origin resource | enabling a link to trigger a download which may not be immediately apparent to the user | CSP |
| Server hosting linked resource | someone downloads a file rather than browses to it | `Content-Disposition` (and/or CORS?) |

Currently, Chrome, Firefox, and Safari have all opted to mitigate the second risk. I may be missing something, but it seems like a way to protect someone from, say, downloading the source of `index.php` from a remote server. Perhaps there are other use cases as well. Hopefully server-side devs are protecting their sensitive files in other ways as well. It seems like CORS (to your earlier point) could be used in this scenario as well, to enable only sanctioned domains access to the contents of particular directories (should it ever be necessary to permit downloads in the first place).

As for the other side of the equation, there is some risk to end users if files are downloaded in the background (which this kind of link enables). Ignoring the `download` attribute on all cross-domain links in the way these browsers have certainly helps mitigate this risk in most scenarios, but it does not give any control to the authors of the primary domain to override this restriction in any way. And there are instances where it would be useful in terms of UX.

Hopefully this will clear up why I’m suggesting a feature like this:

Consider a site like Twitter. When viewing an image posted in a tweet (which is CORS-approved, embedded in the HTML), having an explicit download link—rather than relying on context menus, which may or may not be available, depending on the UA & device—could be useful. They host their media on a separate domain, so these links (in the aforementioned browsers) would not operate as a user might expect.

To enable the desired experience as simply as possible, the media server would currently need to support two URIs for addressing the same resource: one for displaying the resource (e.g., `https://foo.com/img.png`) and one for downloading it (e.g., `https://foo.com/img.png?download`) and the server would need to be configured to add a `Content-Disposition` header.

If there was a means by which a site like Twitter’s could explicitly enable downloads from their media host (which seems like the realm of CSP), it seems like it would solve the issue of cross-domain downloads while still mitigating the two risks *and* enable the use of a single URI for a single resource rather than multiple (which I believe was at least partly the purpose of adding the `download` attribute in the first place). Perhaps they would need to set up complementary CORS rules, but I imagine they are already in place because the images are already embedded on their other domains.

**Comment by annevk** [comment]
Sorry, it's still unclear to me why the primary site declaring something beyond `<a download href={cross-origin-url}>` is a useful signal to anyone. Would the browser behave differently with or without the CSP?

It seems to me that the primary site is the attacker in all scenarios and an attacker might as well use CSP.

**Comment by annevk** [comment]
See https://bugzilla.mozilla.org/show_bug.cgi?id=676619#c5.
